plainkey/keydata.go

```diff
@@ -0,0 +1,264 @@
+)
+
+// hashAlg corresponds to a digest algorithm.
+// XXX: This is the third place this appears now - we almost certainly want to put this
+// in one place. Maybe for another PR.
+type hashAlg crypto.Hash
+
+func (a hashAlg) Available() bool {
+	return crypto.Hash(a).Available()
+}
+
+func (a hashAlg) New() hash.Hash {
+	return crypto.Hash(a).New()
+}
+
+func (a hashAlg) Size() int {
+	return crypto.Hash(a).Size()
+}
+
+func (a hashAlg) MarshalASN1(b *cryptobyte.Builder) {
+	b.AddASN1(cryptobyte_asn1.SEQUENCE, func(b *cryptobyte.Builder) { // AlgorithmIdentifier ::= SEQUENCE {
+		var oid asn1.ObjectIdentifier
+
+		switch crypto.Hash(a) {
+		case crypto.SHA1:
+			oid = sha1Oid
+		case crypto.SHA224:
+			oid = sha224Oid
+		case crypto.SHA256:
+			oid = sha256Oid
+		case crypto.SHA384:
+			oid = sha384Oid
+		case crypto.SHA512:
+			oid = sha512Oid
+		default:
+			b.SetError(fmt.Errorf("unknown hash algorithm: %v", crypto.Hash(a)))
+			return
+		}
+		b.AddASN1ObjectIdentifier(oid) // algorithm OBJECT IDENTIFIER
+		b.AddASN1NULL()                // parameters ANY DEFINED BY algorithm OPTIONAL
+	})
+}
+
+func (a hashAlg) MarshalJSON() ([]byte, error) {
+	var s string
+
+	switch crypto.Hash(a) {
+	case crypto.SHA1:
+		s = "sha1"
+	case crypto.SHA224:
+		s = "sha224"
+	case crypto.SHA256:
+		s = "sha256"
+	case crypto.SHA384:
+		s = "sha384"
+	case crypto.SHA512:
+		s = "sha512"
+	case crypto.Hash(nilHash):
+		s = "null"
+	default:
+		return nil, fmt.Errorf("unknown hash algorithm: %v", crypto.Hash(a))
+	}
+
+	return json.Marshal(s)
+}
+
+func (a *hashAlg) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	switch s {
+	case "sha1":
+		*a = hashAlg(crypto.SHA1)
+	case "sha224":
+		*a = hashAlg(crypto.SHA224)
+	case "sha256":
+		*a = hashAlg(crypto.SHA256)
+	case "sha384":
+		*a = hashAlg(crypto.SHA384)
+	case "sha512":
+		*a = hashAlg(crypto.SHA512)
+	default:
+		// be permissive here and allow everything to be
+		// unmarshalled.
+		*a = nilHash
+	}
+
+	return nil
+}
+
+func deriveAESKey(ikm, salt []byte) []byte {
+	r := hkdf.New(crypto.SHA256.New, ikm, salt, []byte("ENCRYPT"))
+
+	key := make([]byte, 32)
+	if _, err := io.ReadFull(r, key); err != nil {
+		panic(fmt.Sprintf("cannot derive key: %v", err))
+	}
+
+	return key
+}
+
+type additionalData struct {
+	Version    int
+	Generation int
+	KDFAlg     hashAlg
+	AuthMode   secboot.AuthMode
+}
+
+func (d additionalData) MarshalASN1(b *cryptobyte.Builder) {
+	b.AddASN1(cryptobyte_asn1.SEQUENCE, func(b *cryptobyte.Builder) {
+		b.AddASN1Int64(int64(d.Version))
+		b.AddASN1Int64(int64(d.Generation))
+		d.KDFAlg.MarshalASN1(b)
+		b.AddASN1Enum(int64(d.AuthMode))
+	})
+}
+
+type platformKeyId struct {
+	Alg    hashAlg `json:"alg"`
+	Salt   []byte  `json:"salt"`
+	Digest []byte  `json:"digest"`
+}
+
+type keyData struct {
+	Version int    `json:"version"`
+	Nonce   []byte `json:"nonce"`
+
+	PlatformKeyID platformKeyId `json:"platform-key-id"`
```

Comment:
it would be good to have a doc comment on this one

Reply:
I added a comment here.

I added a comment for the Nonce field as well and then realized it didn't make sense because the field was holding 2 unrelated values. I've pushed another change that splits this into 2 distinct fields.